### PR TITLE
detect/analyzer: add more details for tcp_mss - v6

### PR DESCRIPTION
**Pull request by 0xEniola**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6355

Previous PR: #9777

Describe changes:
- Incorporate feedback from previous PR.

Output:
```
{
  "raw": "alert tcp any any -> any any (msg: "Testing mss"; tcp.mss: 123-456; sid: 4;)",
  "id": 4,
  "gid": 1,
  "rev": 0,
  "msg": "Testing mss",
  "app_proto": "unknown",
  "requirements": [],
  "type": "pkt",
  "flags": [
    "src_any",
    "dst_any",
    "sp_any",
    "dp_any",
    "need_packet",
    "toserver",
    "toclient"
  ],
  "pkt_engines": [
    {
      "name": "packet",
      "is_mpm": false
    }
  ],
  "frame_engines": [],
  "lists": {
    "packet": {
      "matches": [
        {
          "name": "tcp.mss",
          "tcp_mss": {
            "operand": "range",
            "min": 123,
            "max": 456
          }
        }
      ]
    }
  }
}
```

```
{
  "raw": "alert tcp any any -> any any (msg: "Testing mss"; tcp.mss: >=439; sid: 5;)",
  "id": 5,
  "gid": 1,
  "rev": 0,
  "msg": "Testing mss",
  "app_proto": "unknown",
  "requirements": [],
  "type": "pkt",
  "flags": [
    "src_any",
    "dst_any",
    "sp_any",
    "dp_any",
    "need_packet",
    "toserver",
    "toclient"
  ],
  "pkt_engines": [
    {
      "name": "packet",
      "is_mpm": false
    }
  ],
  "frame_engines": [],
  "lists": {
    "packet": {
      "matches": [
        {
          "name": "tcp.mss",
          "tcp_mss": {
            "operand": "greater than or equal to",
            "value": 439
          }
        }
      ]
    }
  }
}
```

```
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1462
```

### CODE REVIEW

**Review by jufajardini — APPROVED** [review]
Looks good to me, thanks! :)

File: src/detect-tcpmss.c [hunk 1]
```diff
@@ -62,6 +62,33 @@ void DetectTcpmssRegister(void)
     return;
 }
 
+/**
+ * \brief Return human readable value for tcp.mss mode
+ *
+ * \param mode uint8_t DetectU16Data tcp.mss mode value
+ */
+const char *DetectTcpmssModeToString(DetectUintMode mode)
```
Comment (catenacyber) [inline src/detect-tcpmss.c:70]: This function should rather be in rust/src/detect/uint.rs as it is generic for integers, not specific to tcp mss, right ?

File: src/detect-engine-analyzer.c [hunk 2]
```diff
@@ -861,6 +863,22 @@ static void DumpMatches(RuleAnalyzer *ctx, JsonBuilder *js, const SigMatchData *
                 jb_close(js);
                 break;
             }
+            case DETECT_TCPMSS: {
+                const DetectU16Data *cd = (const DetectU16Data *)smd->ctx;
+
+                jb_open_object(js, "tcp_mss");
+                const char *flag = DetectTcpmssModeToString(cd->mode);
+                jb_set_string(js, "operand", flag);
+                if (strcmp(flag, "range") == 0) {
```
Comment (catenacyber) [inline src/detect-engine-analyzer.c:872]: I would use the `cd->mode` integer for comparison instead of its derivated string ;-)

Comment (0xEniola) [reply src/detect-engine-analyzer.c:872]: It's what I used at first; I was asked to change it.

So!

File: src/detect-tcpmss.c (hunk 1 above)
Comment (0xEniola) [reply src/detect-tcpmss.c:70]: Better still to be in the `detect-engine-uint.c` file, I think.

Comment (catenacyber) [reply src/detect-tcpmss.c:70]: detect-engine-uint.c is just a wrapper to rust/src/detect/uint.rs

You can use rust/src/detect/uint.rs directly ;-)

File: src/detect-engine-analyzer.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-engine-analyzer.c:872]: Where ? I did not find it in previous PRs...

File: src/detect-tcpmss.c (hunk 1 above)
Comment (0xEniola) [reply src/detect-tcpmss.c:70]: Well, Idk.

I don't know rust, & I just got accustomed to C during this exercise.
So I made sure not to go too deep.

File: src/detect-engine-analyzer.c (hunk 2 above)
Comment (0xEniola) [reply src/detect-engine-analyzer.c:872]: Wait! Are you talking about the “range” operand?

File: src/detect-tcpmss.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-tcpmss.c:70]: Ok, in the end, I think we will need it to be rust

File: src/detect-engine-analyzer.c (hunk 2 above)
Comment (0xEniola) [reply src/detect-engine-analyzer.c:872]: Oh! Since `mode` had been converted to string, doing the comparison with `strcmp` was the only way.

Comment (catenacyber) [reply src/detect-engine-analyzer.c:872]: you still have access to `cd->mode` ;-) (and it is cheaper to compare integer than strings)

Comment (0xEniola) [reply src/detect-engine-analyzer.c:872]: True True.

File: src/detect-tcpmss.c (hunk 1 above)
Comment (0xEniola) [reply src/detect-tcpmss.c:70]: > Ok, in the end, I think we will need it to be rust

Hello, so I was looking through the `uint.rs` source file for this, and I was thinking; "Wouldn't it be better if the function is placed in the `detect-engine-uint.c` file instead?".

**Comment by jufajardini** [comment]
Adding the `decision-required` label here as @victorjulien wants to decide on how we tackle cases of keywords with a `range`.

File: src/detect-engine-analyzer.c [hunk 3]
```diff
@@ -861,6 +863,22 @@ static void DumpMatches(RuleAnalyzer *ctx, JsonBuilder *js, const SigMatchData *
...
+                    jb_set_uint(js, "min", cd->arg1);
+                    jb_set_uint(js, "max", cd->arg2);
```
Comment (victorjulien) [inline src/detect-engine-analyzer.c:874]: @jasonish we need to settle on what to use here for single values and ranges. Any thoughts on what makes sense?

Comment (jasonish) [reply src/detect-engine-analyzer.c:874]: Minimally, with sime inspiration from the Elasticsearch query language:
```json
		{
		    "name": "tcp.mss",
		    "equals": 5
		},
		{
		    "name": "tcp.mss",
		    "range": {
			"min": 5,
			"max": 10
		    }
		}
		{
		    "name": "tcp.mss",
		    "gt": 5
		}
```

But this breaks the format, is there a reason why we use "name", then use a sub-object other than we are already doing it? To keep with the current format I think this looks good:
```json
		{
		    "name": "tcp.mss",
		    "tcp_mss": {
			"equals": 5
		    }
		},
		{
		    "name": "tcp.mss",
		    "tcp_mss": {
			"range": {
			    "min": 5,
			    "max": 10,
			}
		    }
		},
		{
		    "name": "tcp.mss",
		    "tcp_mss": {
                        "operator": "gt",
			"gt": 5
		    }
		},
```
Note that I added the `operator` to the last one, but I don't think its really needed.

File: src/detect-tcpmss.c (hunk 1 above)
Comment (jufajardini) [reply src/detect-tcpmss.c:70]: Really sorry for none of us ever getting to answer these before.

> Will I have to wrap the function into the detect-engine-uint.c file too?

Usually, if we need a Rust function to be available to the C side, we will do that still from the Rust code, (see for instance https://github.com/OISF/suricata/blob/master/rust/src/detect/uint.rs#L303)

> Based on my thoughts, I'd have to write a function that uses the return values of function detect_parse_uint_mode

I don't think that your function would change much from what it looks like today, just that it should be in the Rust side, as Philippe mentioned. You are already using the right Parameter type, and then would still return a string.
Comparing to `detect_parse_uint_mode`,  it looks to me that you would almost have params and return in reverse from what is there :P

**Review by catenacyber — CHANGES_REQUESTED** [review]
Needs a rebase, and there are comments to address

File: src/detect-engine-analyzer.c (hunk 3 above)
Comment (victorjulien) [reply src/detect-engine-analyzer.c:874]: I'm good with Jason's last suggestion here.

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/10824

**Pull request closed without merging**